### ISSUE TITLE
Reconsider Rakudo build structure

### ISSUE THREAD

**Issue by vrurg** [perl6/problem-solving#94]
Current Rakudo build process still suffers from early single-core times when nobody was thinking about new language releases yet. Most notable problem is `6.c` core being a special case in many locations: from Makefile to language version support code.

**Comment by vrurg** [comment]
Proposed solution:

1. `core` directory is getting renamed into `core.c`
1. `CORE.setting` becomes `CORE.c.setting`.
1. `BOOTSTRAP.nqp` and `EXPORTHOW.nqp` are migrating into release-specific subdirs.
1. `Perl6::BOOTSTRAP` module becomes `Perl6::BOOTSTRAP::v6<rev>`, similar to `CORE::v6<rev>` pseudo-package.
1. Revisions `d` and `e` receive their own `Perl6::BOOTSTRAP::v6<rev>` modules.

The last two items are needed to support proper overriding of revision cores. I have stumbled upon a problem with `Grammar` class which clearly demonstrates the issue. `BOOTSTRAP.nqp` sets default parent and `GrammarHOW` for the class. But it does so exclusively for `Grammar` for revision `c`. The one defined for `e` remains with a usual `ClassHOW`. Besides, class code refers to `HLL::Compiler` which is only available if core is loading `Perl6::BOOTSTRAP`. `use NQPHLL:from<NQP>` doesn't work due to problems with serialization (sorry, didn't go deeper into this direction). For this reason `e`-specific `BOOTSTRAP` is the best solution for the problem. 

This separation would also allow for things like the one proposed by @jnthn : redefinition of `EXPORTHOW` with revision-specific classes; and use of revision-specific `HOW`s. The latter actually means that `Metamodel` might get it's own `.<rev>` subdirectories.

**Comment by vrurg** [comment]
Reference implementation is in rakudo/rakudo#3138.

**Comment by jnthn** [comment]
> core directory is getting renamed into core.c
> CORE.setting becomes CORE.c.setting.

These seem reasonable.

> BOOTSTRAP.nqp and EXPORTHOW.nqp are migrating into release-specific subdirs.
> Perl6::BOOTSTRAP module becomes Perl6::BOOTSTRAP::v6<rev>, similar to CORE::v6<rev> pseudo-package.
> Revisions d and e receive their own Perl6::BOOTSTRAP::v6<rev> modules.

How do these relate to each other? For example, the increasing core versions have a lexical nesting relationship; I'm guessing that's now how it is in this case, but what is it?

**Comment by vrurg** [comment]
> > BOOTSTRAP.nqp and EXPORTHOW.nqp are migrating into release-specific subdirs.
> > Perl6::BOOTSTRAP module becomes Perl6::BOOTSTRAP::v6, similar to CORE::v6 pseudo-package.
> > Revisions d and e receive their own Perl6::BOOTSTRAP::v6 modules.
> 
> How do these relate to each other? For example, the increasing core versions have a lexical nesting relationship; I'm guessing that's now how it is in this case, but what is it?

No, there is no relation between them. It makes no sense. Each bootstrap is only bound to its corresponding `CORE` which kind of `use` it (`World.nqp`, line 768). The only purpose of `Perl6::BOOTSTRAP::v6e` for now, for example, is to provide `HLL::Compiler` for CORE.e.

**Comment by jnthn** [comment]
OK, that sounds sensible. Thanks. Aside from my point about the curious versioning of the `NULL` "setting" on the PR (I think there should be just one NULL, which really means "no setting"), I'm fine with this.
